nixos/modules/services/system/nix-daemon-firewall.nix

```diff
@@ -0,0 +1,142 @@
+{ config, lib, ... }:
+let
+
+  cfg = config.nix.firewall;
+
+in
+{
+  options.nix.firewall = {
+    enable = lib.mkEnableOption "firewalling for outgoing traffic of the nix daemon";
+
+    allowLoopback = lib.mkOption {
+      description = "Whether not not to allow traffic on the loopback interface. Traffic is still subject to protocol/port rules";
+      default = false;
+      example = true;
+    };
+
+    allowPrivateNetworks = lib.mkOption {
+      description = "Whether not not to allow traffic to local networks. Traffic is still subject to protocol/port rules. Note that this option may break DNS resolution when the DNS resolver is in a local network";
+      default = true;
+      example = false;
+    };
+
+    allowNonTCPUDP = lib.mkOption {
+      description = "Whether to allow traffic that is neither TCP nor UDP";
+      type = lib.types.bool;
+      default = false;
+      example = true;
+    };
+
+    allowedTCPPorts = lib.mkOption {
+      description = "TCP ports to which traffic is allowed. Specifying no ports will allow all TCP traffic";
+      type = lib.types.listOf (
+        lib.types.oneOf [
+          lib.types.singleLineStr
+          lib.types.port
+        ]
+      );
+      default = [ ];
+      example = [
+        "http"
+        443
+        "30000-31000"
+      ];
+    };
+
+    allowedUDPPorts = lib.mkOption {
+      description = "UDP ports to which traffic is allowed. Specifying no ports will allow all UDP traffic";
+      type = lib.types.listOf (
+        lib.types.oneOf [
+          lib.types.singleLineStr
+          lib.types.port
+        ]
+      );
+      default = [ ];
+      example = [ 53 ];
+    };
+
+    extraNftablesRules = lib.mkOption {
+      description = "Extra nftables rules to prepend to the generated ones";
+      type = lib.types.listOf lib.types.singleLineStr;
+      default = [ ];
+      example = [ "ip daddr 1.1.1.1 udp dport accept" ];
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    # Ensure we can properly use nftables
+    assertions = [
+      {
+        assertion = config.networking.nftables.enable;
+        message = ''
+          The nix-daemon firewall requires an nftables-based firewall.
+          networking.nftables.enable must be set to true.
+        '';
+      }
+      {
+        assertion = !config.networking.nftables.flushRuleset;
+        message = ''
+          The nix-daemon firewall writes extra tables to nftables.
+          networking.nftables.flushRuleset must be set to false.
+        '';
+      }
+    ];
+
+    systemd.services.nix-daemon = {
+      after = [ "nftables.service" ];
+      # Add the cgroup ID to a nft set on daemon start
+      serviceConfig.NFTSet = "cgroup:inet:nix_daemon_firewall:nix_daemon";
+    };
+
+    # Generate nftables rules
+    networking.nftables.checkRuleset = false; # XXX: Does not work for now, not sure why
```

Comment:
I think this is likely because we use lkl for checking ruleset and it uses a fairly old linux kernel that might not support all nftables features that you are using here.

Reply:
I found an option: [671b0b7](https://github.com/NixOS/nixpkgs/pull/464613/commits/671b0b7ea6569ecc3e188d35c3061d1298b0a06c)